src/main/java/org/opensearch/plugin/insights/core/service/categorizer/QueryShapeGenerator.java

```diff
@@ -76,6 +76,17 @@ public class QueryShapeGenerator {
     static final Map<Class<?>, List<Function<Object, String>>> AGG_FIELD_DATA_MAP = FieldDataMapHelper.getAggFieldDataMap();
     static final Map<Class<?>, List<Function<Object, String>>> SORT_FIELD_DATA_MAP = FieldDataMapHelper.getSortFieldDataMap();
 
+    /**
+     * Method to get query shape hash code given a source
+     * @param source search request source
+     * @param showFields whether to include field data in query shape
+     * @return Hash code of query shape as Integer
+     */
+    public static int getShapeHashCode(SearchSourceBuilder source, Boolean showFields) {
+        String shape = buildShape(source, showFields);
+        return shape.hashCode();
```

Comment:
String hashcode is the simplest option and suitable for most use cases. It is used in Java objects such as HashMap, HashSet... That being said it is 32-bit.

If we are concerned about collisions, I would recommend a non-cryptographic option since they are faster and we do not care about security.

OpenSearch core has a 128-bit [MurmurHash3](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/common/hash/MurmurHash3.java) library which we can use. I did a quick mockup:

```
    public static int getShapeHashCode(SearchSourceBuilder source, Boolean showFields) {
        String shape = buildShape(source, showFields);

        final BytesRef shapeBytes = new BytesRef(shape);
        return MurmurHash3.hash128(shapeBytes.bytes, shapeBytes.offset, shapeBytes.length, 0, SEED).hashCode();
    }
```

The [Hash128.hashcode()](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/common/hash/MurmurHash3.java#L71) method converts the Hash128 object back to a 32-bit int so becomes no better than String hashcode. Depending on how we use the hash, it may not be necessary or practical to go beyond 64 bits. In that case we could use half of Hash128.

Reply:
Thanks for the details, my vote would be for `MurmurHash3` and 64 bit should probably be enough for our use-cases

Reply:
Updated method to use Murmur lib and return 64-bit long

Reply:
@dzane17 - I am wondering which bytes are trimmed off during lossy conversion from 128 bytes to 64 bytes. And if that can have any impact on the number of collisions. Also, does the hashcode need to be numeric? Can it not be a string? cc: @deshsidd

Reply:
@jainankitk We are taking `hash128.h1` which javadoc defines as "lower 64 bits part" so higher 64 bits are trimmed. Taking top or bottom half bytes should not matter and I saw this method used at least once in the OpenSearch codebase.
```
        return MurmurHash3.hash128(shapeBytes.bytes, 0, shapeBytes.length, 0, new MurmurHash3.Hash128()).h1;
```

This will reduce the hash space but we agreed 2^64 is a good compromise.

Reply:
In the future we may need to keep track of the hash->shapeString relation with something like `HashMap<String, String>` or `HashMap<Long, String>`. However, HashMap calls the hashcode() method of the key type. Both  String.hashcode() and Long.hashcode() return an int so that would reduce our hash space back to 32 bits.

I think we will need a custom Map implementation or OpenSearch has a [LongLongHash.java](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/common/util/LongLongHash.java) class that we could possible use with 128 bit keys. In either case, numeric keys should be easier to work with than Strings.

Reply:
It can be string too if required.

Reply:
> In the future we may need to keep track of the hash->shapeString relation with something like HashMap<String, String> or HashMap<Long, String>. However, HashMap calls the hashcode() method of the key type. Both String.hashcode() and Long.hashcode() return an int so that would reduce our hash space back to 32 bits.

The reduction in hash space is not really a concern as it will be resolved using collision chain within the HashMap. We won't be losing the unique query shape hashcode altogether, which can be the case during lossy conversion from 128 bit murmur encoding to 64 bit. We can also consider using `Pair<Long,Long>` for storing this 128 bit query hashcode

Reply:
I have changed return type to `MurmurHash3.Hash128` which is an object that contains two longs. `MurmurHash3.Hash128` already has hashcode() and equals() methods implemented so we can use it directly as HashMap keys in the future (`HashMap<MurmurHash3.Hash128, String>`).

Reply:
Nice!